glean-core/src/storage/mod.rs

```diff
@@ -31,14 +31,30 @@ impl StorageManager {
         store_name: &str,
         clear_store: bool,
     ) -> Option<JsonValue> {
-        let mut snapshot: HashMap<&str, HashMap<String, JsonValue>> = HashMap::new();
+        let mut snapshot: HashMap<String, HashMap<String, JsonValue>> = HashMap::new();
 
         let mut snapshotter = |metric_name: &[u8], metric: &Metric| {
-            let map = snapshot
-                .entry(metric.category())
-                .or_insert_with(HashMap::new);
             let metric_name = String::from_utf8_lossy(metric_name).into_owned();
-            map.insert(metric_name, metric.as_json());
+            if metric_name.contains("/") {
+                let category = format!("labeled_{}", metric.category());
+                let map = snapshot.entry(category).or_insert_with(HashMap::new);
+
+                let mut s = metric_name.splitn(2, '/');
+                let metric_name = s.next().expect("No metric name");
+                let label = s.next().expect("No metric label");
+
+                let obj = map.entry(metric_name.into()).or_insert_with(|| json!({}));
+                if let Some(obj) = obj.as_object_mut() {
```

Comment:
Won't this always be the case, given the `or_insert_with` above?

Reply:
In theory yes. Handling it here shields us against broken data (where we _don't_ get an object for some reason .... wait, now that I'm writing this down we're constructing this object completely on our own so we provide the guarantees on what's in there)